File: 0234-palindrome-linked-list/0234-palindrome-linked-list.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def isPalindrome(self, head: Optional[ListNode]) -> bool:
        reverse = None
        start = head
        while start!=None:
            t = ListNode(start.val)
            t.next = reverse
            reverse = t
            start = start.next
        
        while reverse != None and head != None:
            if reverse.val != head.val:return False
            reverse = reverse.next
            head = head.next
        return True

File: 0234-palindrome-linked-list/0234-palindrome-linked-list.cpp

class Solution {
public:
    ListNode* reverse(ListNode* head) {
        ListNode* prev = nullptr;
        ListNode* curr = head;

        while (curr != nullptr) {
            ListNode* nextNode = curr->next;
            curr->next = prev;
            prev = curr;
            curr = nextNode;
        }

        return prev;
    }

    bool isPalindrome(ListNode* head) {
        if (head == nullptr || head->next == nullptr)
            return true;

        ListNode* slow = head;
        ListNode* fast = head;

        while (fast->next != nullptr && fast->next->next != nullptr) {
            slow = slow->next;
            fast = fast->next->next;
        }

        ListNode* secondHalf = reverse(slow->next);

        ListNode* firstHalf = head;
        ListNode* temp = secondHalf;

        while (temp != nullptr) {
            if (firstHalf->val != temp->val) {
                reverse(secondHalf);
                return false;
            }
            firstHalf = firstHalf->next;
            temp = temp->next;
        }
        slow->next = reverse(secondHalf);

        return true;
    }
};